RIOT-OS/RIOT
# cpu/esp32: support for light/deep sleep and pm_layered

# Contribution description

This PR adds support for EPS32's light and deep sleep power modes and make them available by `pm_layered`.

The power management of ESP32 supports now the following operating modes:

- **Modem-sleep**: default operating mode when the WiFi interface is disabled.
- **Active** mode: default operating mode when the WiFi interface is used
- **Light-sleep** mode: the CPUs including peripherals are stalled, the SRAM is retained
- **Deep-sleep** CPU and the SRAM are powered down, the RTC memory can be retained

Since the peripherals are not working during light-sleep and deep-sleep, the CPU cannot be woken up by internal interrupt sources such as timers. Therefore, RIOT's layered power management can't select them as idle mode. The application has to select them explicitly using the `pm_set` function.

To **exit from light-sleep**, the following wake-up sources can be used:
- RTC timer (set the RTC timer alarm using `rtc_set` before calling `pm_set`)
- GPIOs that are configured as input with enabled interrupt
- RxD signal of UART0 and/or UART1 (configured with `ESP_PM_WUP_UART0` and `ESP_PM_WUP_UART1`)

**Please note** Since only level interrupts are supported in light-sleep mode, defined edge interrupts of type `GPIO_RISING` and `GPIO_FALLING` are implicitly mapped to `GPIO_HIGH` and `GPIO_LOW`, respectively, when entering Light-sleep mode.

To **exit from deep-sleep**, only RTC resources can be used as wake-up sources:
- RTC timer (set the RTC timer alarm using `rtc_set` before calling `pm_set`)
- RTC GPIOs (configured by `ESP_PM_WUP_PINS` and `ESP_PM_WUP_LEVEL`)

In _Deep-sleep_ mode the SRAM is powered down. However, the slow RTC memory can be retained. Therefore, data that must be retained during _Deep-sleep_ and the subsequent system restart have to be stored in the slow RTC memory. For that purpose, use
- `__attribute__((section(".rtc.bss")))` to place uninitialized data in section `.rtc.bss`, and
- `__attribute__((section(".rtc.data")))` to place initialized data in section `.rtc.data`.

### Testing procedure

Use a ESP32 board with a button at GPIO0 (usually default) and flash `tests/periph_pm`, for example:
```
CFLAGS='-DESP_PM_WUP_UART0=6 -DESP_PM_WUP_PINS=GPIO0 -DESP_PM_WUP_LEVEL=0' \
BOARD=esp32-wroom-32 make -C tests/periph_pm flash
```
This configures the RxD signal of UART0 as wake-up source from light-sleep and the button at GPIO0 as wake-up source from light and deep sleep.

Connect the ESP32 board via a USB multimeter and perform the following tests while observing the values on the USB multimeter. (**Please note**: Although the ESP32 consumes only about 800 uA in light sleep mode and only 10 uA in deep sleep mode, the USB multimeter can measure values between 10 mA and 34 mA. The reason for this is that there are a number of on-board components, such as the USB2UART bridge, that generate a base load. For the test it is sufficient that in light sleep mode approx. 21 mA less is consumed and in deep sleep approx. 22 mA less than in normal operating mode.)

- [ ] Unblock the light-sleep (1) mode for 5 seconds:
  ```
  unblock_rtc 1 5
  ```
  ESP32 should go into light sleep mode (1) for 5 seconds and then continue working correctly.

- [ ] Unblock the deep-sleep mode (0) for 15 seconds and afterwards unblock the light-sleep (1) mode:
  ```
  unblock_rtc 0 15
  unblock 1
  ```
  Once the light-sleep mode (1) is unblocked by the second command, ESP32 should go into the deep-sleep mode for the remaining time. The system should restart with cause `rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)` after that.

- [ ] Set light-sleep mode (1), wait a bit and press _Enter_ at the console
  ```
  set 1
  ```
  The system should return from light-sleep once _Enter_ was pressed.

- [ ] Set deep-sleep mode (0), wait a bit and press the button at GPIO0.
  ```
  set 0
  ```
  The system should restart with cause `rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)` once the button was pressed.

### Issues/PRs references

Fixes issue #13365 
~PR #13417 should be merged before this PR to ensure consistent documentation.~
Depends on PR #13655
